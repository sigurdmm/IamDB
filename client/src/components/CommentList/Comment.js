import PropTypes from 'prop-types';

import { dateToHuman } from '../../utils/dateUtil';
import './Comment.less';

const Comment = ({ createdAt, children }) => <div className="comment">
    <strong className="comment__published">
      Published: <em>{dateToHuman(new Date(parseInt(createdAt, 10)))}</em>
    </strong>
    <p className="comment__text">{children}</p>
  </div>;
Comment.propTypes = {
  createdAt: PropTypes.any,
  children: PropTypes.string,
};

export default Comment;
