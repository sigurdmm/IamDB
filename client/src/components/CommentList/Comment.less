@import '../../constants';

.comment {
  background-color: @white;
  box-shadow: rgba(0,0,0,.05) 1px 2px 4px;
  box-sizing: border-box;

  padding: 0.5em;
  border-top: @primaryColor solid 1px;

  &__published {
    color: @primaryColor;
  }

  &__text {
    font-size: 0.8em;
  }
}
