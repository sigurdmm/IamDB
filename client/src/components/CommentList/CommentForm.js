import {
  Formik,
  Form,
  Field,
  ErrorMessage,
} from 'formik';
import React from 'react';
import PropTypes from 'prop-types';

import './CommentForm.less';

const CommentForm = ({ onSubmit }) => <Formik
  initialValues={{ comment: '' }}
  onSubmit={({ comment }) => onSubmit(comment)}
  validate={(values) => {
    const errors = {};

    if (!values.comment || values.comment.length < 10) {
      errors.comment = `Comment cannot be less than 10 characters. Your's is ${values.comment.length}`;
    }

    return errors;
  }}>
  <Form className="commentform">
    <label htmlFor="comment" className="commentform__label">Comment</label>
    <Field
      className="commentform__comment"
      id="comment"
      name="comment"
      rows="10"
      component="textarea"/>
    <ErrorMessage className="error" name="comment" component="div"/>

    <button type="submit">Submit comment</button>
  </Form>
</Formik>;

CommentForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
};

export default CommentForm;
