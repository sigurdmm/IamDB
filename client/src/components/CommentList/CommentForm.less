
.commentform {
  box-sizing: border-box;

  padding: 0.5em 0;

  &__label {
    display: block;
  }

  &__comment {
    box-sizing: border-box;
    display: block;

    margin: 1em 0;

    width: 100%;
  }

  button[type="submit"] {
    display: block;

    padding: 0.5em 1em;
  }
}
