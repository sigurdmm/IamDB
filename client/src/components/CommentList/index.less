@import '../../constants';

.commentlist {

  &__list {
    list-style-type: none;

    padding: 0;
  }

  &__list > li~li{
    margin-top: 1em;
  }
}
