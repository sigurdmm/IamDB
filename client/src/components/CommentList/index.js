import PropTypes from 'prop-types';

import CommentForm from './CommentForm';
import Comment from './Comment';

import './index.less';

const CommentList = ({ comments, onSubmit }) => <div className="commentlist">
  <ul className="commentlist__list">
    {comments.map((comment, idx) => <li key={`comment-${idx}`}>
      <Comment createdAt={comment.createdAt}>{comment.text}</Comment>
    </li>)}
  </ul>
  <CommentForm onSubmit={onSubmit}/>
</div>;

CommentList.propTypes = {
  comments: PropTypes.arrayOf(PropTypes.shape({
    createdAt: PropTypes.any,
    text: PropTypes.string.isRequired,
  })),
  onSubmit: PropTypes.func.isRequired,
};

export default CommentList;
