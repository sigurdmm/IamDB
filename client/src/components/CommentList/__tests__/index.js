import { render } from 'enzyme';

import CommentList from '../index';

const comments = [
  {
    createdAt: '1542296465964',
    text: 'Hello world',
  },
  {
    createdAt: '1542296465974',
    text: 'Hello world 2',
  },
];

describe('<CommentList/>', () => {
  it('should render without comments', () => {
    const wrapper = render(<CommentList comments={[]} onSubmit={() => null}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render list of comments', () => {
    const wrapper = render(<CommentList comments={comments} onSubmit={() => null}/>);

    expect(wrapper).toMatchSnapshot();
  });
});
