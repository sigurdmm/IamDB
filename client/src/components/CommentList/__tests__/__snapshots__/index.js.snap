// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentList/> should render list of comments 1`] = `
<div
  class="commentlist"
>
  <ul
    class="commentlist__list"
  >
    <li>
      <div
        class="comment"
      >
        <strong
          class="comment__published"
        >
          Published: 
          <em>
            15/10/2018
          </em>
        </strong>
        <p
          class="comment__text"
        >
          Hello world
        </p>
      </div>
    </li>
    <li>
      <div
        class="comment"
      >
        <strong
          class="comment__published"
        >
          Published: 
          <em>
            15/10/2018
          </em>
        </strong>
        <p
          class="comment__text"
        >
          Hello world 2
        </p>
      </div>
    </li>
  </ul>
  <form
    class="commentform"
  >
    <label
      class="commentform__label"
      for="comment"
    >
      Comment
    </label>
    <textarea
      class="commentform__comment"
      id="comment"
      name="comment"
      rows="10"
    />
    <button
      type="submit"
    >
      Submit comment
    </button>
  </form>
</div>
`;

exports[`<CommentList/> should render without comments 1`] = `
<div
  class="commentlist"
>
  <ul
    class="commentlist__list"
  />
  <form
    class="commentform"
  >
    <label
      class="commentform__label"
      for="comment"
    >
      Comment
    </label>
    <textarea
      class="commentform__comment"
      id="comment"
      name="comment"
      rows="10"
    />
    <button
      type="submit"
    >
      Submit comment
    </button>
  </form>
</div>
`;
